File: pkg/material/mirror.go
package material

import (
	"math/rand"

	"sif/gogs/eric/relray/pkg/geometry"
	"sif/gogs/eric/relray/pkg/spectrum"
	"sif/gogs/eric/relray/pkg/vec"
)

// Mirror is a perfect specular reflector with spectral reflectance.
type Mirror struct {
	Reflectance spectrum.SPD
}

func (m *Mirror) Scatter(inDir vec.Vec3, hit geometry.Hit, rng *rand.Rand) ScatterResult {
	reflected := inDir.Reflect(hit.Normal)
	return ScatterResult{
		Scattered:   true,
		OutDir:      reflected.Normalize(),
		Reflectance: m.Reflectance,
	}
}

func (m *Mirror) Emitted(hit geometry.Hit) spectrum.SPD {
	return spectrum.SPD{}
}

File: rust/src/material/mirror.rs
use crate::geometry::Hit;
use crate::spectrum::Spd;
use crate::vec::Vec3;
use rand::rngs::SmallRng;

use super::{Material, ScatterResult};

/// Perfect specular reflector with spectral reflectance.
pub struct Mirror {
    pub reflectance: Spd,
}

impl Material for Mirror {
    fn scatter(&self, in_dir: Vec3, hit: &Hit, _rng: &mut SmallRng) -> ScatterResult {
        let reflected = in_dir.reflect(hit.normal);
        ScatterResult {
            scattered: true,
            out_dir: reflected.normalize(),
            reflectance: self.reflectance,
        }
    }

    fn emitted(&self, _hit: &Hit) -> Spd {
        Spd::default()
    }
}
